feat: add C implementation for `math/base/special/gamma-lanczos-sum`

gunjjoshi opened the pull request:
Resolves #1822

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [`math/base/special/gamma-lanczos-sum`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/gamma-lanczos-sum).

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #649 
-   fixes #1822 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   ☑️ Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

raunak-dev-edu committed:
feat: add `iter/do-until-each`

PR-URL: #1408
Closes: #807

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Pranavchiku committed:
fix: resolve failing addon build in `math/base/special/maxabs`

PR-URL: #1826
Closes:  stdlib-js/todo#2113

gunjjoshi committed:
feat: add C implementation for math/base/special/gamma-lanczos-sum

gunjjoshi committed:
feat: add C implementation for math/base/special/gamma-lanczos-sum

gunjjoshi committed:
feat: add C implementation for math/base/special/gamma-lanczos-sum

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/include/stdlib/math/base/special/gamma_lanczos_sum.h:
```diff
@@ -0,0 +1,41 @@
+#endif
+
+/**
+* Calculates the Lanczos sum approximation.
```
```suggestion
* Calculates the Lanczos sum for the approximation of the gamma function.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/lib/native.js:
```diff
@@ -0,0 +1,67 @@
+// MAIN //
+
+/**
+* Calculates the Lanczos sum approximation.
```
```suggestion
* Calculates the Lanczos sum for the approximation of the gamma function.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/lib/native.js:
```diff
@@ -0,0 +1,67 @@
+/**
+* Calculates the Lanczos sum approximation.
+*
+* @name gammaLanczosSum
+* @type {Function}
```
```suggestion
* @private
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/src/main.c:
```diff
@@ -0,0 +1,74 @@
+#include "stdlib/constants/float64/pinf.h"
+
+/**
+* Calculates the Lanczos sum approximation.
```
```suggestion
* Calculates the Lanczos sum for the approximation of the gamma function.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/src/main.c:
```diff
@@ -0,0 +1,74 @@
+/**
+* Calculates the Lanczos sum approximation.
+*
+* @param {number} x - input value
+* @returns {number} Lanczos sum approximation
```
```suggestion
* @param x    input value
* @return     Lanczos sum approximation
```

We use separate markup for C implementations.

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/src/main.c:
```diff
@@ -0,0 +1,74 @@
+		ax = x;
+	}
+	if ( ax <= 1.0 ) {
+		s1 = 38474670393.31777 + (x * (36857665043.51951 + (x * (15889202453.72942 + (x * (4059208354.298835 + (x * (680547661.1834733 + (x * (78239755.00312005 + (x * (6246580.776401795 + (x * (341986.3488721347 + (x * (12287.194511824551 + (x * (261.61404416416684 + (x * 2.5066282746310007))))))))))))))))))); // eslint-disable-line max-len
```
1. We don't lint C with ESLint.
2. You've inlined the evalrational implementation. We don't want to do this, as now the coefficients are hardcoded. Instead, update the script in the scripts directory to inject an `evalrational` function, as done in other packages. This function then becomes a one-liner calling that evalrational function. A decent C compiler will handle inlining.

kgryte reviewed (changes requested):
@gunjjoshi Thanks for working on this. Left a comment regarding the implementation. Once the implementation is updated using code generation, this should be on the path to merging.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

Snehil-Shah committed:
feat: add C implementation for `math/base/special/acsch`

Closes: https://github.com/stdlib-js/stdlib/issues/860
PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1833
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>

Planeshifter committed:
build: run C benchmarks in CI workflow

Userf1995 committed:
feat: add `math/base/assert/is-nonnegative-finite`

PR-URL: #1829
Closes: #1828

saisrikardumpeti committed:
feat: add `math/base/special/asecd`

PR-URL: #1788
Closes: #41

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Jai0401 committed:
docs: improve README examples for `stats/base/dists/rayleigh` namespace

PR-URL: #1754
Closes: #1642

---------

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

aman-095 committed:
feat: add C implementation for `math/base/special/log2`

PR-URL: #1717

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
docs: improve README examples of `stats/base/dists/negative-binomial`

PR-URL: #1774
Closes: #1637

---------

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update list of contributors

PR-URL: #1837

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update related packages sections 

PR-URL: #1838
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

gunjjoshi committed:
fix: updated script to inject evalrational function

gunjjoshi committed:
fix: updated script to inject evalrational function

gunjjoshi commented:
Thanks for pointing this out @kgryte. I have modified `scripts/evalrational.js`. Now, it injects the required functions in `main.c`. Am I correct in the implementation now ? Any other suggestions ?

gunjjoshi committed:
feat: add C implementation for `math/base/special/gammasgn`

PR-URL: #1841
Closes: #649

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/ahaversin`

PR-URL: #1840
Closes: #1420

---------

Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

performant23 committed:
feat: add C implementation for `math/base/special/xlogy`

PR-URL: #1827
Closes: #1815

---------

Signed-off-by: Rutam <138517416+performant23@users.noreply.github.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/cscd`

PR-URL: #1786
Closes: #39

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/acotd`

PR-URL: #1785
Closes: #43

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/acscd`

PR-URL: #1789
Closes: #42

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/secd`

PR-URL: #1787
Closes: #38

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update REPL namespace documentation

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1843
Reviewed-by: Athan Reines <kgryte@gmail.com>
Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/acosd`

PR-URL: #1782 
Closes: #36

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/atand`

PR-URL: #1784
Closes: #37

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
build: delete gypfile field when files are not present

saisrikardumpeti committed:
feat: add `math/base/special/asind`

PR-URL: #1781
Closes: #35

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
chore: update package meta data

PR-URL: #1856

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
docs: fix JSDoc description in `math/base/special/acoversin`

PR-URL: #1853

Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

adityacodes30 committed:
feat: add `string/base/last-code-point`

PR-URL: #1755
Closes: #1702

---------

Signed-off-by: Aditya Sapra <110766802+adityacodes30@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update related packages sections

PR-URL: #1857

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

kgryte committed:
docs: update contributors

kgryte committed:
fix: update dependencies

kgryte committed:
fix: use correct package names

kgryte committed:
fix: update include path

kgryte committed:
fix: update build configurations

performant23 committed:
refactor: update `blas/ext/base/drev` to follow current project conventions

Closes: https://github.com/stdlib-js/stdlib/issues/1488
PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1839
Ref: https://github.com/stdlib-js/stdlib/issues/1152
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Pranav Goswami <goswami.4@iitj.ac.in>
Signed-off-by: Athan Reines <kgryte@gmail.com>
Co-authored-by: Pranav Goswami <goswami.4@iitj.ac.in>

kgryte committed:
fix: update build configurations

kgryte committed:
fix: update build configurations

kgryte committed:
fix: update build configurations

kgryte committed:
fix: update build configurations and fix argument extraction

kgryte committed:
fix: update build configurations

stdlib-bot committed:
docs: update REPL namespace documentation

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1858
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/bernoulli`

PR-URL: #1860 
Closes: #1664

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

AuenKr committed:
refactor: update `blas/ext/base/dssum` to follow current projects conventions

PR-URL: #1861
Closes: #1501

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

Userf1995 committed:
feat: add C implementation for `math/base/assert/int32-is-odd`

PR-URL: #1865
Closes: #1806 

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

aman-095 committed:
feat: add C implementation for `math/base/assert/is-safe-integer`

PR-URL: #1862

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

performant23 committed:
feat: add C implementation for `math/base/assert/is-nonpositive-finite`

PR-URL: #1859 
Closes: #1831

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update REPL namespace documentation

PR-URL: #1877

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/xlog1py`

PR-URL: #1847 
Closes: #1842

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com> 
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

gunjjoshi committed:
Update main.c

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/scripts/evalrational.js:
```diff
@@ -46,6 +51,19 @@ var P = [
 	261.6140441641668190791708576058805625502,
 	2.506628274631000502415573855452633787834
 ];
+var P1 = [
```
Why is `P1` necessary here? Why doesn't `P` suffice?

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/scripts/evalrational.js:
```diff
@@ -24,10 +24,15 @@
 var currentYear = require( '@stdlib/time/current-year' );
 var licenseHeader = require( '@stdlib/_tools/licenses/header' );
 var compile = require( '@stdlib/math/base/tools/evalrational-compile' );
+var compileC = require( '@stdlib/math/base/tools/evalpoly-compile-c' );
```
Ah, I see. You need to use `evalrational-compile-c`. This will remove the need for `P1` and `Q1`, etc. You'll need to update here and in the C file.

gunjjoshi replied:
Thanks for pointing this out, I will make the suggested changes.

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/src/main.c:
```diff
@@ -0,0 +1,128 @@
+#include "stdlib/math/base/special/gamma_lanczos_sum.h"
+#include "stdlib/constants/float64/pinf.h"
+
+// BEGIN: rational_pq_p
```
These need to be updated based on the changes to the script.

kgryte commented on lib/node_modules/@stdlib/math/base/special/gamma-lanczos-sum/src/main.c:
```diff
@@ -0,0 +1,128 @@
+* // returns ~950.366
+*/
+double stdlib_base_gamma_lanczos_sum( const double x ) {
+	double ax;
```
This will also need to be updated once `evalrational-compile-c` is used.

kgryte replied:
@gunjjoshi Was this function auto-generated by the script or did you manually copy? I'm a little bit surprised that the script understood to use the constant macro at L43.

gunjjoshi replied:
Hey @kgryte, the function is indeed auto-generated. But the only modification I did was to replace the macro. Is there any other way for handling macros here ?

kgryte replied:
Manually editing is risky, as, if we ever update the constants and rerun the script, your manual edits will be overwritten. My thinking is that we need to update https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/tools/evalrational-compile-c/lib/main.js in order to handle this particular edge case. Namely, if the "ratio" is infinity, then inline `1.0 / 0.0` (or `-1.0 / 0.0`).

kgryte replied:
In particular, we need to update `value2string` https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/tools/evalrational-compile-c/lib/main.js#L53. Would you like to open a separate PR for this? Would also need to add associated tests.

gunjjoshi replied:
Yes @kgryte, I would be opening a separate PR regarding this, making necessary changes in the `value2string` function.

kgryte replied:
Great! Thanks!

gunjjoshi replied:
Now since we changed `evalrational-compile-c`, I ran the script again, to auto-generate the function. It auto-generated the functions. It passes all the tests and benchmarks well.

But, while pushing the changes, it gives the following error:

```bash
lib/node_modules/@stdlib/random/base/minstd-shuffle/README.md
  78:1-87:4  error  9:1   error   Displayed return value is `20739838`, but expected `1421600654` instead   stdlib/doctest  eslint  remark-lint

✖ 1 error
make: *** [markdown-examples-javascript-files] Error 1

Encountered an error when running README JavaScript examples.

error: failed to push some refs to 'https://github.com/gunjjoshi/stdlib'
```

The [package](https://github.com/stdlib-js/stdlib/blob/cf29ed0fc4f7108b5be01daca89ef72e6ebfcc1f/lib/node_modules/%40stdlib/random/base/minstd-shuffle/README.md?plain=1#L86) that leads to the error is very much unrelated to what we are doing here. I wonder what might be the reason for this.

cc: @kgryte

kgryte replied:
This happened when attempting to rebase, I am guessing?

kgryte replied:
For this one time, you can circumvent Git hooks by appending `--no-verify` when running the `push` command.

kgryte replied:
I also addressed the lint error. If you rebase on the latest `develop`, you should be able to push as you would normally.

kgryte replied:
*sigh* Looks like I should have recommended just merging in the latest develop, rather than `rebase`. No worries. Looks like GitHub is still picking up on the changes in this PR and we'll squash-merge so all good.

gunjjoshi replied:
Yes, it happened while rebasing. But after getting the latest develop, it works well. Thanks !
Also, in the latest commit pushed, I've regenerated the auto-generated function, and it now works well with infinities, as we intended.

gunjjoshi committed:
feat: add C implementation for `math/base/special/boxcox`

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1870
Ref: https://github.com/stdlib-js/stdlib/issues/649
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/boxcoxinv`

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1878
Ref: https://github.com/stdlib-js/stdlib/issues/649
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>

gunjjoshi committed:
refactor: used evalrational-compile-c

gunjjoshi committed:
refactor: used evalrational-compile-c

gunjjoshi committed:
refactor: used evalrational-compile-c

gunjjoshi commented:
Hey @kgryte I have made the changes suggested by you, by using `evalrational-compile-c`. Is this the correct implementation ?

gunjjoshi committed:
feat: add C implementation for `math/base/special/boxcox1p`

PR-URL: #1883

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/boxcox1pinv`

PR-URL: #1885

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

performant23 committed:
feat: add C implementation for `math/base/assert/uint32-is-pow2`

PR-URL: #1886
Closes: #1868

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
chore: remove indentation

Planeshifter committed:
build: add workflow to auto-assign labels to PRs

Planeshifter committed:
build: rename labeler config file

Userf1995 committed:
docs: improve README examples of `stats/base/dists/truncated-normal`

PR-URL: #1872
Cloes: #1647

---------

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
build: explicitly set token used in action

saisrikardumpeti committed:
feat: add `math/base/special/tand`

PR-URL: #1771
Closes: #34

---------

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/cotd`

PR-URL: #1772
Closes: #40

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

kgryte committed:
fix: remove labels

The previous configuration was generating false positives.

aman-095 committed:
feat: add C implementation for `math/base/assert/is-prime`

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1866
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Athan Reines <kgryte@gmail.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/acotd`

PR-URL: #1879 
Closes: #649

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/asecd`

PR-URL: #1880

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/acsc`

PR-URL: #1887
Closes: #859

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/atand`

PR-URL: #1881

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/acosd`

PR-URL: #1871

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

performant23 committed:
feat: add C implementation for `math/base/special/factorial2`

PR-URL: #1898 
Closes: #1892

---------

Signed-off-by: Rutam <138517416+performant23@users.noreply.github.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

kgryte committed:
docs: clean-up examples and descriptions

Ref: #1339

kgryte committed:
feat: add `countSameValue` to namespace

kgryte committed:
test: add test case to ensure full test coverage

Ref: #1339
Ref: https://github.com/stdlib-js/stdlib/commit/634044b1b369b08b2748c3f6f457c2b1259ccfa3

kgryte committed:
docs: add notes concerning algorithm

Ref: #1339
Ref: https://github.com/stdlib-js/stdlib/commit/634044b1b369b08b2748c3f6f457c2b1259ccfa3

kgryte committed:
docs: add same value example

Ref: #1339
Ref: https://github.com/stdlib-js/stdlib/commit/634044b1b369b08b2748c3f6f457c2b1259ccfa3

kgryte committed:
docs: clean-up documentation and examples and rename folder

Ref: #1384
Ref: https://github.com/stdlib-js/stdlib/commit/1ed81b78fd6469d374c7b033f258e709934539b2

kgryte committed:
feat: add `countSameValueZero` to namespace

kgryte committed:
docs: fix copy

stdlib-bot committed:
docs: update related packages sections 

PR-URL: #1907

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

kgryte committed:
docs: update test messages, comments, and benchmarks

Ref: #1354
Ref: https://github.com/stdlib-js/stdlib/commit/609682c8ef64ecf0fb6a4d7094e226a99a985ab5

kgryte committed:
feat: add `isNonNegativeFinite` to namespace

kgryte committed:
feat: add `isNonNegativeFinite` to namespace

gunjjoshi committed:
feat: add C implementation for `math/base/special/fast/hypot` 

PR-URL: #1910

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/fast/acosh`

PR-URL: #1897

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Co-authored-by:  Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/fast/abs`

PR-URL: #1894

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update namespace table of contents

PR-URL: #1920

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/asind`

PR-URL: #1876

---------

Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update namespace table of contents

PR-URL: #1921

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com> 
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/fast/asinh`

PR-URL: #1908

---------

Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
fix: move addition by one into the sqrt invocation

Planeshifter committed:
build: update check given leading stdlib prefix is already included

Planeshifter committed:
docs: clean-up C function parameter and return annotations

Planeshifter committed:
docs: fix function names in C example code

stdlib-bot committed:
chore: update package meta data

PR-URL: #1926

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Athan Reines <kgryte@gmail>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

AuenKr committed:
refactor: update `blas/ext/base/dnansum` to follow current project conventions

PR-URL: #1899
Closes: #1483

---------

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Reviewed-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

stdlib-bot committed:
docs: update REPL namespace documentation

PR-URL: #1928

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: Planeshifter <1913638+Planeshifter@users.noreply.github.com>
Reviewed-by: Planeshifter <1913638+Planeshifter@users.noreply.github.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/fast/max`

PR-URL: #1912

---------

Signed-off-by: Pranav Goswami <goswami.4@iitj.ac.in>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav Goswami <goswami.4@iitj.ac.in>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/fast/min`

PR-URL: #1913
Closes: #1911

---------

Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Signed-off-by: Pranav Goswami <goswami.4@iitj.ac.in>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav Goswami <goswami.4@iitj.ac.in>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update related packages sections

PR-URL: #1930

---------

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

performant23 committed:
fix: update variable type declarations in `math/base/assert/int32-is-odd`

PR-URL: #1935

Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>

Rec0iL99 committed:
feat: add `utils/any-own-by`

PR-URL: #1851
Closes: #819

---------

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
feat: update namespace TypeScript declarations

PR-URL: #1937

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Reviewed-by: Planeshifter <1913638+Planeshifter@users.noreply.github.com>

gunjjoshi committed:
feat: add C implementation for `math/base/special/modf`

PR-URL:  #1934
Closes: #1916

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav Goswami <goswami.4@iitj.ac.in>
Co-authored-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>

gunjjoshi committed:
feat: add C implementation for `math/base/special/fast/pow-int`

PR-URL: #1919
Closes: #1917

---------

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Pranav Goswami <goswami.4@iitj.ac.in>
Co-authored-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

saisrikardumpeti committed:
feat: add `math/base/special/cosd`

PR-URL: #1777
Closes: #33

---------

Signed-off-by: Sai Srikar Dumpeti <80447788+the-r3aper7@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Userf1995 committed:
feat: add C implementation for `math/base/special/fast/atanh`

PR-URL: #1909

---------

Signed-off-by: Rejoan Sardar <119718513+Rejoan-Sardar@users.noreply.github.com>
Signed-off-by: Pranav Goswami <goswami.4@iitj.ac.in>
Co-authored-by:Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

performant23 committed:
feat: add C implementation for `math/base/assert/is-composite`

PR-URL: #1918
Closes: #1901

---------

Signed-off-by: Rutam <138517416+performant23@users.noreply.github.com>
Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
Co-authored-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>
Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
build: update installed R version

R linting seems to be broken on CI, see `lint_random_files.yml` workflow
run such as the following:

https://github.com/stdlib-js/stdlib/actions/runs/8319437589/job/22762740425#step:15:988

This PR updates the installed R version to see if that resolves the
issue as all dependencies are correctly installed from CRAN locally
when ussing that version.

Planeshifter committed:
build: skip labeler for auto PRs and only assign when all files match

Planeshifter committed:
build: update R lint configuration

stdlib-bot committed:
docs: update list of contributors

PR-URL: #1940

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Reviewed-by: Planeshifter <1913638+Planeshifter@users.noreply.github.com>

stdlib-bot committed:
docs: update related packages sections

PR-URL: #1941

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

AuenKr committed:
refactor: update `blas/ext/base/dcusum` to follow current project conventions

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1867
Closes: https://github.com/stdlib-js/stdlib/issues/1471
Ref: https://github.com/stdlib-js/stdlib/issues/1152
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Athan Reines <kgryte@gmail.com>

stdlib-bot committed:
docs: update REPL namespace documentation

PR-URL: #1956

Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>
Co-authored-by: kgryte <2643044+kgryte@users.noreply.github.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

performant23 committed:
fix: missing headers in C examples

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1958
Reviewed-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Pranav Goswami <goswami.4@iitj.ac.in>

gunjjoshi committed:
Update test.native.js

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
feat: add `constants/float32/half-pi`

PR-URL: 	https://github.com/stdlib-js/stdlib/pull/1953
Ref: https://github.com/stdlib-js/stdlib/issues/1944
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>

kgryte committed:
fix: update documentation to reflect correct seed range

The seed for MT19937 must be a nonzero positive integer, as discussed
in the original paper describing the Mersenne Twister PRNG.

Closes: https://github.com/stdlib-js/stdlib/issues/1963

kgryte committed:
docs: update contributing guide to explicitly require running `make install`

Closes: https://github.com/stdlib-js/stdlib/issues/1678

kgryte committed:
docs: add note regarding spaces in directory path

kgryte committed:
docs: add note concerning native add-ons

kgryte committed:
docs: add link

gunjjoshi committed:
Merge branch 'develop' into gamma-lanczos-sum

gunjjoshi committed:
Merge branch 'develop' into gamma-lanczos-sum

gunjjoshi committed:
feat: add C implementation for math/base/special/gamma-lanczos-sum

gunjjoshi committed:
Merge branch 'develop' into gamma-lanczos-sum

kgryte reviewed (approved):
LGTM. Thanks, @gunjjoshi!
